File: algorithm/sliding_window.go
package algorithm

import (
	"math"
	"sync"
	"time"
)

// SlidingWindowState holds the two fixed-window counters used by the
// sliding-window counter algorithm (weighted estimate of previous + current).
type SlidingWindowState struct {
	mu sync.Mutex

	prevCount int64
	currCount int64

	// windowStart is the beginning of the current fixed window.
	windowStart time.Time
	expiry      time.Time
}

type SlidingWindowConfig struct {
	Limit  int
	Window time.Duration
}

func (cfg *SlidingWindowConfig) Evaluate(state *SlidingWindowState, cost int, now time.Time) Result {
	state.mu.Lock()
	defer state.mu.Unlock()

	cfg.roll(state, now)
	state.expiry = now.Add(cfg.Window * 2)

	fLimit := float64(cfg.Limit)
	elapsed := now.Sub(state.windowStart).Seconds()
	weight := 1 - elapsed/cfg.Window.Seconds()
	if weight < 0 {
		weight = 0
	}
	used := float64(state.currCount) + float64(state.prevCount)*weight

	if used+float64(cost) <= fLimit {
		state.currCount += int64(cost)
		return Result{
			Allowed:   true,
			Remaining: cfg.Limit - int(math.Ceil(used)) - cost,
		}
	}

	// Estimate when enough of the previous window ages out for this request to
	// pass. Conservative: report at least 1 second, at most one window.
	retryAfter := cfg.Window
	if state.prevCount > 0 {
		need := used + float64(cost) - fLimit // excess to shed
		shedPerSec := float64(state.prevCount) / cfg.Window.Seconds()
		if shedPerSec > 0 {
			retryAfter = time.Duration(math.Ceil(need/shedPerSec)) * time.Second
		}
	}
	if retryAfter < time.Second {
		retryAfter = time.Second
	}
	if retryAfter > cfg.Window {
		retryAfter = cfg.Window
	}

	return Result{
		Allowed:    false,
		Remaining:  0,
		RetryAfter: retryAfter,
	}
}

// roll advances the fixed-window counters if `now` has crossed into a new
// window. Must be called with state.mu held.
func (cfg *SlidingWindowConfig) roll(state *SlidingWindowState, now time.Time) {
	if state.windowStart.IsZero() {
		state.windowStart = now.Truncate(cfg.Window)
		return
	}

	elapsedWindows := int64(now.Sub(state.windowStart) / cfg.Window)
	if elapsedWindows <= 0 {
		return
	}
	if elapsedWindows == 1 {
		state.prevCount = state.currCount
	} else {
		// More than a full window has passed: both counters are stale.
		state.prevCount = 0
	}
	state.currCount = 0
	state.windowStart = state.windowStart.Add(time.Duration(elapsedWindows) * cfg.Window)
}

func (s *SlidingWindowState) IsExpired(now time.Time) bool {
	s.mu.Lock()
	defer s.mu.Unlock()
	return !s.expiry.IsZero() && now.After(s.expiry)
}

File: backend/lua/sliding_window.lua
-- Sliding Window Counter (weighted previous + current fixed-window counters)
--
-- KEYS[1]   : rate limit key
-- ARGV[1]   : window size in milliseconds
-- ARGV[2]   : current timestamp (unix ms)
-- ARGV[3]   : limit (max requests per window)
-- ARGV[4]   : cost (requests to consume)
-- ARGV[5]   : TTL in seconds
--
-- Returns {allowed(0|1), remaining, retry_after_seconds}

local key    = KEYS[1]
local window = tonumber(ARGV[1])
local now    = tonumber(ARGV[2])
local limit  = tonumber(ARGV[3])
local cost   = tonumber(ARGV[4])
local ttl    = tonumber(ARGV[5])

-- Current fixed-window start, aligned to the window boundary.
local curr_start = now - (now % window)

local state = redis.call("HMGET", key, "prev", "curr", "start")
local prev  = tonumber(state[1]) or 0
local curr  = tonumber(state[2]) or 0
local start = tonumber(state[3])

if start == nil then
    prev, curr, start = 0, 0, curr_start
elseif curr_start > start then
    -- Advance the fixed windows.
    local elapsed_windows = math.floor((curr_start - start) / window)
    if elapsed_windows == 1 then
        prev = curr
    else
        prev = 0 -- more than a full window passed: both counters are stale
    end
    curr  = 0
    start = curr_start
end

-- Weighted estimate of usage in the sliding window.
local weight = 1 - ((now - start) / window)
if weight < 0 then weight = 0 end
local used = curr + prev * weight

if used + cost <= limit then
    curr = curr + cost
    redis.call("HMSET", key, "prev", prev, "curr", curr, "start", start)
    if ttl > 0 then
        redis.call("EXPIRE", key, ttl)
    end
    local remaining = limit - math.ceil(used) - cost
    if remaining < 0 then remaining = 0 end
    return {1, remaining, 0}
end

-- Estimate when enough of the previous window ages out for this request to pass.
local retry_after = window / 1000.0
if prev > 0 then
    local need       = used + cost - limit
    local shed_per_s = prev / (window / 1000.0)
    if shed_per_s > 0 then
        retry_after = math.ceil(need / shed_per_s)
    end
end
if retry_after < 1 then retry_after = 1 end
if retry_after > window / 1000.0 then retry_after = math.ceil(window / 1000.0) end

return {0, 0, retry_after}
